scripts/sct_pipeline.py

```diff
@@ -317,37 +325,53 @@ def run_function(function, folder_dataset, list_subj, list_args=[], nb_cpu=None,
     # add full path to each subject
     list_subj_path = [os.path.join(folder_dataset, subject) for subject in list_subj]
 
-    # All scripts that are using multithreading with ITK must not use it when using multiprocessing on several subjects
+    # All scripts that are using multithreading with ITK must not use it when using multiprocessing
     os.environ["ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS"] = "1"
 
     # create list that finds all the combinations for function + subject path + arguments. Example of one list element:
     # ('sct_propseg', os.path.join(path_sct, 'data', 'sct_test_function', '200_005_s2''), '-i ' + os.path.join("t2", "t2.nii.gz") + ' -c t2')
     list_func_subj_args = list(itertools.product(*[[function], list_subj_path, list_args]))
         # data_and_params = itertools.izip(itertools.repeat(function), data_subjects, itertools.repeat(parameters))
 
-    # Computing Pool for parallel process, distribute2mpi.MpiPool in MPI environment, multiprocessing.Pool otherwise
     sct.log.debug("stating pool with {} thread(s)".format(nb_cpu))
-    pool = Pool(nb_cpu)
+    pool = PoolExecutor(nb_cpu)
     compute_time = None
     try:
         compute_time = time()
-        sct.log.debug('paused but print')
-        async_results = pool.map_async(function_launcher, list_func_subj_args)
-        pool.close()
-        pool.join()  # waiting for all the jobs to be done
+        count = 0
+        all_results = []
+
+        sct.log.info('Waiting for results, be patient')
+        future_dirs = {pool.submit(function_launcher, subject_arg): subject_arg
+                         for subject_arg in list_func_subj_args}
+
+        for future in concurrent.futures.as_completed(future_dirs):
+            count += 1
+            subject = os.path.basename(future_dirs[future][1])
+            arguments = future_dirs[future][2]
+            try:
+                result = future.result()
+            except Exception as exc:
+                sct.log.error('{} {} generated an exception: {}'.format(subject, arguments, exc))
+
+            sct.log.info('{}/{}: {} done'.format(count, len(list_func_subj_args), subject))
+            all_results.append(result)
         compute_time = time() - compute_time
-        all_results = async_results.get()
+
         # concatenate all_results into single Panda structure
-        results_dataframe = pd.concat([result for result in all_results])
+        results_dataframe = pd.concat(all_results)
+
     except KeyboardInterrupt:
         sct.log.warning("\nCaught KeyboardInterrupt, terminating workers")
-        pool.terminate()
-        pool.join()
+        pool.shutdown()
```

Comment:
`shutdown()` is only called upon exception here, however it should be called even when there is no exception right ? Otherwise all processes and management thread will be kept alive.

A `finally` exception clause would be better here since this code is also being repeated.